quicklogic/qlf_k4n8/utils/repacker/repack.py

```diff
@@ -0,0 +1,1394 @@
+    leaf blocks by copying them.
+    """
+
+    def walk(block):
+
+        # This is a leaf
+        if block.is_leaf and not block.is_open:
+
+            # Find matching cell
+            cell = eblif.find_cell(block.name)
+            assert cell is not None, block
+
+            # Copy attributes and parameters
+            block.attributes = dict(cell.attributes)
+            block.parameters = dict(cell.parameters)
+
+        # Recurse
+        else:
+            for child in block.blocks.values():
+                walk(child)
+
+    # Walk over CLBs
+    for block in packed_netlist.blocks.values():
+        walk(block)
+
+
+# =============================================================================
+
+
+def load_repacking_rules(json_root):
+    """
+    Loads rules for the repacker from a parsed JSON file
+    """
+
+    # Get the appropriate section
+    json_rules = json_root.get("repacking_rules", None)
+    assert json_rules is not None
+    assert isinstance(json_rules, list), type(json_rules)
+
+    # Convert the rules
+    logging.debug(" Repacking rules:")
+
+    rules = []
+    for entry in json_rules:
+        assert isinstance(entry, dict), type(entry)
+
+        rule = RepackingRule(
+            src=entry["src_pbtype"],
+            dst=entry["dst_pbtype"],
+            index_map=entry["index_map"],
+            port_map=entry["port_map"],
+            mode_bits=entry["mode_bits"],
+        )
+        rules.append(rule)
+
+        logging.debug("  {} -> {}".format(rule.src, rule.dst))
+
+    return rules
+
+
+def expand_port_maps(rules, clb_pbtypes):
+    """
+    Expands port maps of repacking rules so that they explicitly specify
+    port pins.
+    """
+
+    for rule in rules:
+
+        # Get src and dst pb_types
+        path = [PathNode.from_string(p) for p in rule.src.split(".")]
+        path = [PathNode(p.name, mode=p.mode) for p in path]
+        src_pbtype = clb_pbtypes[path[0].name].find(path)
+        assert src_pbtype, ".".join([str(p) for p in path])
+
+        path = [PathNode.from_string(p) for p in rule.dst.split(".")]
+        path = [PathNode(p.name, mode=p.mode) for p in path]
+        dst_pbtype = clb_pbtypes[path[0].name].find(path)
+        assert dst_pbtype, ".".join([str(p) for p in path])
+
+        # Expand port map
+        port_map = {}
+        for src_port, dst_port in rule.port_map.items():
+
+            # Get pin lists
+            src_pins = list(src_pbtype.yield_port_pins(src_port))
+            dst_pins = list(dst_pbtype.yield_port_pins(dst_port))
+
+            assert len(src_pins) == len(dst_pins), (src_pins, dst_pins)
+
+            # Update port map
+            for src_pin, dst_pin in zip(src_pins, dst_pins):
+                port_map[src_pin] = dst_pin
+
+        rule.port_map = port_map
+
+        # DEBUG
+#        logging.debug(" ", rule.src, "->", rule.dst)
+#        for k, v in rule.port_map.items():
+#            logging.debug("  ", k, "->", v)
```

Comment:
Maybe remove these or get them under an `if DEBUG` block with a global DEBUG variable to set it on or off.